### DISCUSSION TITLE
AI tuning based on the selected difficulty level

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#8270] [General]
As we continue improving the AI, it becomes really challenging, especially for less experienced players. 
"Easy" mode should be really friendly for all novice players, kids or just players, who want to play a chill game against nonirritant enemies.
I would like to ask your opinion on how could we adjust the AI according to difficulty level, @ihhub , @oleg-derevenetz , @zenseii , @Districh-ru , @idshibanov  ?
Maybe you can propose some ideas, what AI aspects could be adjusted, so "easy" mode would become easier, but harder difficulties still introduce the level of toughness our advanced AI is designed for.

**Comment by Branikolog** [comment]
I remember we adjusted TP DD spells for that purpose, but I think more logic should be adapted to difficulty level as it is still not really easy on lower difficulties.
We can adjust https://github.com/ihhub/fheroes2/pull/8256 for this purposes, so on easy difficulty AI would be more lazy or scared of human heroes, if it's not already implemented in that PR for sure. :)

**Comment by LeHerosInconnu** [comment]
Hello everyone,

My thoughts on the subject.
I don't think the AI should be limited in "normal" mode.
The game only needs one "easy" mode.
The "normal" mode should be the mode where the AI is fully operational (without restriction) and with the same operation for resources as the human player.
For "hard", "expert" and "impossible" modes, the AI player has bonuses in starting resources and resource production, and the human player has maluses in starting resources (and possibly in resource production with the extra setting).

Some scenarios in the original game are much harder with the new fheroes2 AI, and some are easier (for example, the fheroes2 AI doesn't have visibility of the entire adventure map at the start of the scenario, which makes it less aggressive).

Since fheroes2 is going to have its own scenario editor, I think it's more appropriate to start from a "clean" base that doesn't take into account the change brought about by the new fheroes2 AI on the old Heroes 2 scenarios.
These old scenarios can still be played in "easy" mode, or in "hard" mode to adjust the difficulty level to that of the original game.
In addition, users will be able to update (or repliquate) these maps for the new fheroes2 AI with the new scenario editor.

In general, in any game, the user expects the "normal" mode to be one in which neither the player nor the AI has any bonuses or maluses, and the AI to be at its optimal level (with no reduced capability).

**Reply by zenseii** [reply]
@LeHerosInconnu 
I second the sentiment about having a difficulty level where neither the human nor computer player has any advantages nor disadvantages. Currently that is missing. I do believe we need at least two slightly easier difficulty levels, and so that has to be Easy and Normal with the current difficulty setup (the bishop chess piece doesn't have a graphic btw...).

I think @oleg-derevenetz's current PR will be very interesting to watch w/r to the AI. https://github.com/ihhub/fheroes2/pull/8273

**Reply by oleg-derevenetz** [reply]
Hi @zenseii

> Currently that is missing.

Normal difficulty is currently the difficulty that could be called as such, almost: it has scouting bonus of 1 and AI will not split weaker stack by one on this difficulty, what else prevents it to be named "where neither the human nor computer player has any advantages nor disadvantages"? Well, it also has the limit of 2 Dimension Door casts per AI hero per turn, but I do not know if it is worth be taken into account or not.

**Reply by zenseii** [reply]
@oleg-derevenetz, yes, these were the limits I had in mind when I wrote that the Normal difficulty wasn't a level playing field, but it's also the principle of not having such limitations and knowing that there are none. It is the closest you could get at simulating a match against a human player, currently.

> it also has the limit of 2 Dimension Door casts per AI hero per turn, but I do not know if it is worth be taken into account or not.

Dimension Door is maybe the most powerful spell, though perhaps the AI isn't using it optimally right now, but, I suppose that could happen in the future. 

I think that we shouldn't be afraid to admit that the fheroes2 AI is challenging enough that playing against it when there are no advantages or disadvantages on either side, is to play on Hard difficulty, not Normal.

My impression from looking at the difficulty code is that the difficulties aren't even well-defined and especially those harder than Normal are very hard to distinguish, especially regarding what sort of experience to expect from them and why a player should pick one over the other. Hard vs Expert, what is the difference? Still just the one scouting radius, gold equivalent of 1 less hero on turn 1 and a couple of resources here and there, and a bunch of other inconsequential differences.

The point is that the overall design of the difficulty levels, their aims, target audiences, differences etc. , should be reviewed and get clearly defined, in my opinion.

**Reply by oleg-derevenetz** [reply]
@zenseii

> My impression from looking at the difficulty code is that the difficulties aren't even well-defined and especially those harder than Normal are very hard to distinguish, especially regarding what sort of experience to expect from them and why a player should pick one over the other. Hard vs Expert, what is the difference? Still just the one scouting radius, gold equivalent of 1 less hero on turn 1 and a couple of resources here and there, and a bunch of other inconsequential differences.

Well, in the original game the differences were mostly in resource bonuses in the beginning of the game and per turn, plus some "unfair" advantages like extra movement points, maybe having the whole map revealed, etc.

> The point is that the overall design of the difficulty levels, their aims, target audiences, differences etc. , should be reviewed and get clearly defined, in my opinion.

Well, you are free to offer some difficulty level hierarchy with specific bonuses or limitations for each level (the simpler the better) :)

**Reply by LeHerosInconnu** [reply]
Hello everyone,

With fheroes2's new AI, players will have to get used to the AI's improved performance.
This is the new "normal" fheroes2 standard. :)

Is there a place where all the bonuses/maluses applied to the human player and the AI according to difficulty are clearly listed?
And what are the current bonuses for AI personalities (builder, warrior, explorer)?

**Reply by oleg-derevenetz** [reply]
> Is there a place where all the bonuses/maluses applied to the human player and the AI according to difficulty are clearly listed?

Not yet, it still has to be settled.

> And what are the current bonuses for AI personalities (builder, warrior, explorer)?

These personalities are just a show. Actually, there is no any personality-dependent behavior, all AI players behave in the same "dynamic" manner, depending on the specific observed circumstances.

**Reply by LeHerosInconnu** [reply]
@oleg-derevenetz.

> > Is there a place where all the bonuses/maluses applied to the human player and the AI according to difficulty are clearly listed?
> 
> Not yet, it still has to be settled.

Difficult to think about proposals without knowing what is currently being applied.

> > And what are the current bonuses for AI personalities (builder, warrior, explorer)?
> 
> These personalities are just a show. Actually, there is no any personality-dependent behavior, all AI players behave in the same "dynamic" manner, depending on the specific observed circumstances.

Too bad there's no difference actually.
In the original game this gave clues as to which opponent to deal with first according to his suggested development method.

**Reply by oleg-derevenetz** [reply]
> Too bad there's no difference actually.
In the original game this gave clues as to which opponent to deal with first according to his suggested development method.

In fheroes2, individual heroes have "personalities". Each AI player has "scouts", "fighters", "couriers", and these roles are re-elected each turn depending on the situation.

**Comment by Branikolog** [comment]
A nice surrender logic was introduced recently. I was so surprised, when AI was defending his castle and was about to be demolished by my forces and pass all his artifacts but just simply surrendered leaving me without a reward for the siege. :)

What about prohibiting this unfair mechanics for easy (not sure about normal, but could be) difficulty? :D

**Reply by zenseii** [reply]
I'm not against defining and potentially restricting the AI behavior according to the different levels. I've seen this in other Heroes games and it felt natural to the game and like every increase in difficulty meant that you were playing against a better player that had other tricks up his sleeve. I can imagine the hit n run tactic be such a behavior.

> this unfair mechanics for easy (not sure about normal, but could be) difficulty? :D

I get what you mean but I'm not sure I would call this an "unfair" mechanic when it is simply how the game works 😅

**Reply by Branikolog** [reply]
@zenseii 

>I get what you mean but I'm not sure I would call this an "unfair" mechanic when it is simply how the game works 😅

Have you ever met such situation in other HoMM games, when AI surrenders with the troops you were about to demolish? Especially while sitting in any castle or town. :) The first time I experienced that I was sooo surprised. :)
